python/pythonmonkey/require.py

```diff
@@ -82,8 +82,11 @@
 globalThis.python.stderr.write = lambda s: sys.stderr.write(s)
 globalThis.python.stdout.read = lambda n: sys.stdout.read(n)
 globalThis.python.stderr.read = lambda n: sys.stderr.read(n)
-globalThis.python.eval = eval
-globalThis.python.exec = exec
+# Python 3.13 dramatically changed how the namespace in `exec`/`eval` works
+# See https://docs.python.org/3.13/whatsnew/3.13.html#defined-mutation-semantics-for-locals
+_locals = {}  # keep the local variables inside `eval`/`exec` to a dict
+globalThis.python.eval = lambda x: eval(x, None, _locals)
+globalThis.python.exec = lambda x: exec(x, None, _locals)
 globalThis.python.getenv = os.getenv
```

Comment:
https://docs.python.org/3/library/functions.html#exec
> In all cases, if the optional parts are omitted, the code is executed in the current scope. If only globals is provided, it must be a dictionary (and not a subclass of dictionary), which will be used for both the global and the local variables. If globals and locals are given, they are used for the global and local variables, respectively. If provided, locals can be any mapping object. Remember that at the module level, globals and locals are the same dictionary.

https://github.com/python/cpython/blob/v3.9.20/Python/bltinmodule.c#L979-L985
```c
if (globals == Py_None) {
        globals = PyEval_GetGlobals();
        if (locals == Py_None) {
            locals = PyEval_GetLocals();
            if (locals == NULL)
                return NULL;
        }
```

Reply:
On `main`

In your first example, `a` is defined in the global scope, `exec` can catch it and re-assign it without any problem.

However, in your second example, `b` is defined in the local scope of the `foo()` function. Since the `exec` is executed inside of `require.py`, it can only see the variables defined in `require.py` as its "locals".

Reply:
"Without the above change, pythonmonkey would lose the ability to mutate parent scope variables with pm.eval. "

Question: is there a problem with losing that ability?

From a clean design POV, it's probably best to not let the languages mutate each other's scopes, except when those scoping objects are deliberately made available to this, i.e. in the case of `globalThis`.

The fact that Python modules have a scoping semantic not understand in JS-land underscores this from my POV.

If we made `python.eval` work more like indirect eval in JS, it would probably be a saner coding environment for our users, even though it might be a breaking change.  And since the only code we really care about at this early stage is our test cases and BiFrost2, we control everything, so we can fix API and code in lock-step if necessary.

Reply:
i.e. to clarify: if we can change PythonMonkey to work with 3.13, and make it work the same in less than 3.13, I think that's probably okay.  If a JS developer REALLY needs to mutate a Python scope, that dev can just pass in the scoping object, right?